checker_c#/N0t3b00kChecker/Db/CheckerDb.cs

```diff
@@ -0,0 +1,53 @@
+﻿using EnoCore.Checker;
+using Microsoft.Extensions.Logging;
+using MongoDB.Driver;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace N0t3b00kChecker.Db
+{
+    public class CheckerDb
+    {
+        private readonly ILogger<CheckerDb> logger;
+        private readonly IMongoCollection<NotebookUser> users;
+        private readonly InsertOneOptions insertOneOptions = new() { BypassDocumentValidation = false };
+
+        public CheckerDb(ILogger<CheckerDb> logger)
+        {
+            this.logger = logger;
+            this.logger.LogDebug($"CheckerDb() via {MongoConnection})");
+            var mongo = new MongoClient(MongoConnection);
+            var db = mongo.GetDatabase("N0t3b00kCheckerDatabase");
+            this.users = db.GetCollection<NotebookUser>("Users");
+            this.users.Indexes.CreateOne(new CreateIndexModel<NotebookUser>(Builders<NotebookUser>.IndexKeys
+                .Ascending(nu => nu.TaskChainId)));
+        }
+
+        public static string MongoHost => Environment.GetEnvironmentVariable("MONGO_HOST") ?? "localhost";
+
+        public static string MongoPort => Environment.GetEnvironmentVariable("MONGO_PORT") ?? "localhost";
+
+        public static string MongoConnection => $"mongodb://{MongoHost}:{MongoPort}";
+
+        public async Task InsertNotebookUser(NotebookUser user, CancellationToken token)
+        {
+            await this.users.InsertOneAsync(user, this.insertOneOptions, token);
+        }
+
+        public async Task<NotebookUser> FindNotebookUser(string taskChainId, CancellationToken token)
+        {
+            var cursor = await this.users.FindAsync(u => u.TaskChainId == taskChainId, cancellationToken: token);
+            var user = await cursor.FirstOrDefaultAsync(token);
+            if (user == null)
+            {
+                throw new MumbleException("Could not find old user");

```

Comment:
No way to differentiate between a fault on our side and a mumble service

Reply:
how could the function potentially know this though?

Reply:
This happens only when the putflag failed, so you will see a putflag checker error if a putflag checker error occurred.

Reply:
I mean, it's wrong. But still I'm closing this as it's pointless to argue <3

Reply:
I am very certain putflag checker errors cause a checker error, don't worry

Reply:
What if putflag never ran? Db got cleared? Getflag got scheduled too early? So many colorful ways this can break.

Reply:
> how could the function potentially know this though?

Store `putflag_failed = true` on a failed putflag run, throw `MumbleException` here, if `putflag_failed`, else `EnoException` / `InternalError`

Reply:
>What if putflag never ran?

If the checker refuses to take the request, it is considered a checker error.

>Db got cleared?

If you consider someone with db access a malicious actor, you will never get this safe.

>Getflag got scheduled too early?

Hardly the checker's responsibility imo. If you drive a car against a tree, it is not the cars fault if it breaks :P

>Store putflag_failed = true on a failed putflag run, throw MumbleException here, if putflag_failed, else EnoException / InternalError

That yields no objective benefit. "putflag checkererror -> team is checkererror" holds without it, and if the data is missing for successful runs, enochecker_test will scream at you.

Reply:
Point is, if it's not the team's fault, it shouldn't result in Mumble. But as I said, no point in discussing.

Reply:
But it is the team's fault if the putflag fails unless it is a checker error, and if it is a checker error the team is considered checker error

Reply:
It's the team's fault if, and only if, the prior putflag failed, and nothing else went wrong (and even then they couldn't do anything to fix this getflag). For all other possible reasons it's our fault.

Reply:
Yeah, and that happens only if the putflag checkererrors (thus yielding the appropriate checkererror) or the checker fails to save the credentials (yielding enochecker_test errors if the user does not save them appropriately and a checkererror if the db is down).

>and even then they couldn't do anything to fix this getflag

They just have to get their service operational. If the putflag fails gracefully after the flag was deposited, the corresponding getflags find it.